sky/skylet/providers/ibm/node_provider.py

```diff
@@ -0,0 +1,924 @@
+                            f"{node['id']} is pending for {pending_time}")
+                        if pending_time > PENDING_TIMEOUT:
+                            logger.error(
+                                f"pending timeout {PENDING_TIMEOUT} reached, "
+                                f"deleting instance {node['id']}"
+                            )
+                            # we won't try to restart a failed node even if
+                            self._delete_node(node["id"])
+                            # avoid adding the node to cached_nodes
+                            continue
+                    else:
+                        self.pending_nodes.pop(node["id"], None)
+
+            # skypilot-adjustment: verify ip is attached to worker nodes as well
+            nic_id = node["network_interfaces"][0]["id"]
+
+            # pylint: disable=line-too-long
+            res = self.ibm_vpc_client.list_instance_network_interface_floating_ips(
+                node["id"], nic_id
+            ).get_result()
+
+            floating_ips = res["floating_ips"]
+            if len(floating_ips) == 0:
+                # not adding a node that's missing floating ip
+                continue
+            else:
+                node["floating_ips"] = floating_ips
+
+            res_nodes.append(node)
+
+        for node in res_nodes:
+            self.cached_nodes[node["id"]] = node
+
+        return [node["id"] for node in res_nodes]
+
+    def is_running(self, node_id) -> bool:
+        """returns whether a node is in status running"""
+        with self.lock:
+            node = self._get_cached_node(node_id)
+            logger.debug(
+                f"""node: {node_id} is_running?
+                {node["status"] == "running"}""")
+            return node["status"] == "running"
+
+    def is_terminated(self, node_id) -> bool:
+        """returns True if a node is either not recorded or
+         not in any valid status."""
+        with self.lock:
+            try:
+                node = self._get_cached_node(node_id)
+                logger.debug(
+                    f"""node: {node_id} is_terminated?
+                    {node["status"] not in
+                    ["running", "starting", "pending"]}""")
+                return node["status"] not in ["running", "starting", "pending"]
+            # pylint: disable=W0703
+            except Exception:
+                return True
+
+    def node_tags(self, node_id) -> Dict[str, str]:
+        """returns tags of specified node id """
+
+        with self.lock:
+            return self.nodes_tags.get(node_id, {})
+
+    def _get_hybrid_ip(self, node_id):
+        """return external ip for head and private ips for workers"""
+
+        # skypilot-adjustment returns external ips for worker nodes as well.
+        node = self._get_cached_node(node_id)
+        fip = node.get("floating_ips")
+        if fip:
+            return fip[0]["address"]
+
+        node = self._get_node(node_id)
```

Comment:
no need to get into it, since i currently removed the function `_get_hybrid_ip` as you suggested on another conversation. 
will reopen this conversation if it turns out there's a need for the function.